### PR TITLE
Fixed up the phpunit config

### PR DESCRIPTION
**Pull request by GrahamCampbell**
We'll need to update the docs to let people know about this.

### CODE REVIEW

**Comment by taylorotwell** [comment]
Can you describe more what was going on and what this fixes?

**Comment by shehi** [comment]
I personally see only cosmetic changes there. He just added suffix for test files which is useful when you have some non-test file there. 👍

**Comment by scrubmx** [comment]
@taylorotwell @shehi I think this fixes https://github.com/laravel/framework/issues/10808

**Comment by shehi** [comment]
@scrubmx Can't be! The cause for that problem can't be THIS SIMPLE! **runs to test it**

**Comment by GrahamCampbell** [comment]
> I personally see only cosmetic changes there.

Not at all!

> Can you describe more what was going on and what this fixes?

Because our routes.php file is not PSR-4 compliant, we can't safely "require" it. PHPUnit's code coverage will attempt to require all files as if they have no "consequences", but our routes file will execute code when you require it. That fact means that when PHPUnit goes through the stages of checking for uncovered files, it will require the routes file directly, thus causing this exception.

NB This error will only happen if your test suite doesn't require the routes file itself at one point. So, if you don't have any tests that boot the routing layer, then you'll see the bug, or, if you ran a subset of your tests that don't load the routes file, you'll see this error.

File: phpunit.xml [hunk 1]
```diff
@@ -10,12 +10,15 @@
...
+            <directory suffix="Test.php">./tests</directory>
         </testsuite>
     </testsuites>
     <filter>
-        <whitelist>
-            <directory suffix=".php">app/</directory>
+        <whitelist processUncoveredFilesFromWhitelist="true">
+            <directory suffix=".php">./app</directory>
+            <exclude>
+                <file>./app/Http/routes.php</file>
```
Comment (GrahamCampbell) [inline phpunit.xml]: this is the really important change - it is NOT cosmetic - it stops phpunit requiring this file when searching for uncovered files

**Comment by GrahamCampbell** [comment]
https://github.com/sebastianbergmann/php-code-coverage/blob/3.3.0/src/CodeCoverage.php#L707

File: phpunit.xml (hunk 1 above)
Comment (shehi) [inline phpunit.xml]: I retract my statement of it being cosmetic :D LOL, thanks for awesome solution.

**Comment by StephenWattbike** [comment]
I still get exactly the same error after updating the config file. I am running phpunit 4.8.2, php 7.0 and Laravel 5.2.

Is there a step I am missing? I have run composer dump-autoload & php artisan clear-compiled etc.

**Comment by GrahamCampbell** [comment]
> Is there a step I am missing?

There might be other files you need to exclude too that you've added.

**Comment by StephenWattbike** [comment]
Thats actually a very good point, thanks for the tip.

**Comment by taylorotwell** [comment]
What error are y'all even talking about?

**Comment by StephenWattbike** [comment]
I am still working on my whitelist, which should fix this error that I get when running phpunit:

PHPUnit 4.8.24 by Sebastian Bergmann and contributors.

Runtime:	PHP 7.0.3-13+deb.sury.org~trusty+1 with Xdebug 2.4.0RC4
Configuration:	/home/vagrant/Code/phpunit.xml

...........                                                       11 / 11 (100%)

Time: 16.74 seconds, Memory: 28.00Mb

PHP Fatal error:  Uncaught Illuminate\Contracts\Container\BindingResolutionException: Target [Illuminate\Contracts\Debug\ExceptionHandler] is not instantiable. in /home/vagrant/Code/bootstrap/cache/compiled.php:1349
Stack trace:
#0 /home/vagrant/Code/bootstrap/cache/compiled.php(1295): Illuminate\Container\Container->build('Illuminate\\Cont...', Array)
#1 /home/vagrant/Code/bootstrap/cache/compiled.php(1857): Illuminate\Container\Container->make('Illuminate\\Cont...', Array)
#2 /home/vagrant/Code/bootstrap/cache/compiled.php(2175): Illuminate\Foundation\Application->make('Illuminate\\Cont...')
#3 /home/vagrant/Code/bootstrap/cache/compiled.php(2144): Illuminate\Foundation\Bootstrap\HandleExceptions->getExceptionHandler()
#4 [internal function]: Illuminate\Foundation\Bootstrap\HandleExceptions->handleException(Object(Symfony\Component\Debug\Exception\FatalThrowableError))
#5 {main}
  thrown in /home/vagrant/Code/bootstrap/cache/compiled.php on line 1349

Fatal error: Uncaught Illuminate\Contracts\Container\BindingResolutionException: Target [Illuminate\Contracts\Debug\ExceptionHandler] is not instantiable. in /home/vagrant/Code/bootstrap/cache/compiled.php on line 1349

Illuminate\Contracts\Container\BindingResolutionException: Target [Illuminate\Contracts\Debug\ExceptionHandler] is not instantiable. in /home/vagrant/Code/bootstrap/cache/compiled.php on line 1349

Call Stack:
   16.7483   27003040   1. Illuminate\Foundation\Bootstrap\HandleExceptions->handleException() /home/vagrant/Code/bootstrap/cache/compiled.php:0
   16.7857   27022976   2. Illuminate\Foundation\Bootstrap\HandleExceptions->getExceptionHandler() /home/vagrant/Code/bootstrap/cache/compiled.php:2144
   16.7857   27022976   3. Illuminate\Foundation\Application->make() /home/vagrant/Code/bootstrap/cache/compiled.php:2175
   16.7857   27023032   4. Illuminate\Container\Container->make() /home/vagrant/Code/bootstrap/cache/compiled.php:1857
   16.7857   27023032   5. Illuminate\Container\Container->build() /home/vagrant/Code/bootstrap/cache/compiled.php:1295

**Comment by shehi** [comment]
@taylorotwell Don't you remember the case when PHPUnit 5 wasn't working in Laravel 5? Well, @GrahamCampbell and team solved that issue, apparently.

**Comment by GrahamCampbell** [comment]
> I am still working on my phpunit.xml processUncoveredFilesFromWhitelist whitelist, which should fix this error that I get when running phpunit. I think its related to this closed issue: laravel/framework#10808

This fixes that closed issue.

**Comment by GrahamCampbell** [comment]
You'll need to add any other files to the whilelist excludes that are of the same nature as routes.php.

**Comment by GrahamCampbell** [comment]
Not doing so is not a bug in laravel, it's just misconfiguration on you part I'm afraid.

**Comment by StephenWattbike** [comment]
So I should use phpunit 5 to avoid this issue altogether? I think @GrahamCampbell said that only phpunit version 4.8 was supported which is why I'm running that version. I will run 5 if I can avoid this whitelisting which may work for a blank project but its becoming painful for me.

I will continue to try the whitelist with 4.8 but I still get the error after excluding my own additions, and other combinations to get it to work. Sorry to drop in this thread I am trying my best to get past this issue.

**Comment by GrahamCampbell** [comment]
> So I should use phpunit 5 to avoid this issue altogether? I think @GrahamCampbell said that only phpunit version 4.8 was supported which is why I'm running that version. I will run 5 if I can avoid this whitelisting which may work for a blank project but its becoming painful for me.

Both are broken.

**Comment by GrahamCampbell** [comment]
Well, either are broken as long as you configure them correctly...

**Comment by GrahamCampbell** [comment]
Well, neither are broken as long as you configure them correctly...

**Comment by StephenWattbike** [comment]
I will continue trying the whitelist and hope your right. Its just strange that I didn't need a whitelist before and now I do. What has changed? also shouldn't the php flag --no-coverage fix this?

**Comment by GrahamCampbell** [comment]
> I didn't need a whitelist before and now I do. What has changed? also shouldn't the php flag --no-coverage fix this?

So long as code coverage uncovered file searching has been enabled, you have always needed this to prevent this from happening, in all versions.

**Comment by StephenWattbike** [comment]
You are probably right and I will try and get it working using the whitelist.. I was running phpunit without any extra flags or extra configuration without an error before I upgraded to Laravel 5.2.  I didn't know about code coverage stuff until I encountered this issue.

**Comment by vlakoff** [comment]
Why targetting master?

**Comment by vlakoff** [comment]
Sorry, I again mixed up these master/develop which differ between laravel/laravel and laravel/framework.

**Comment by taylorotwell** [comment]
Is it possible this is a breaking change at all?

File: phpunit.xml (hunk 1 above)
Comment (crynobone) [inline phpunit.xml]: Shouldn't we just exclude `./app/Http/routes.php` file without introducing other changes (we could make that the new default on 5.3)

Comment (JosephSilber) [inline phpunit.xml]: Why are we so concerned about breaking BC in the app skeleton? Do people actually update their existing apps? How?

Comment (shehi) [inline phpunit.xml]: I have Laravel-5.2 based code (AudithSoftworks/Basis) and I always have 
it up-to-date with the latest Laravel. I update it to the letter, as far 
as I can, unless those updates are not necessary due to my own 
modifications.

Comment (shehi) [inline phpunit.xml]: I'd recommend adding some kind of XML comment into that file, notifying 
users. We have comments everywhere, why not in this file as well?

**Comment by shehi** [comment]
I don't think it can be a breaking change. Noone would go and copy paste 
new code change blindly lol. If they do, it's their problem. Shouldn't 
and Couldn't be able to hinder the progress....

File: phpunit.xml [hunk 2]
```diff
@@ -10,12 +10,15 @@
          stopOnFailure="false">
     <testsuites>
         <testsuite name="Application Test Suite">
-            <directory>./tests/</directory>
+            <directory suffix="Test.php">./tests</directory>
```
Comment (GrahamCampbell) [inline phpunit.xml]: That one is not necessary, but is more useful.

**Comment by GrahamCampbell** [comment]
Nothing made to laravel/laravel is breaking really, like your web middleware changes.

**Comment by StephenWattbike** [comment]
very strange:

if i use:
>./vendor/bin/phpunit

I do not get an error anymore, I got the idea from this stackoverflow:
https://stackoverflow.com/questions/34760614/calling-phpunit-is-not-calling-vendor-bin-phpunit

>PHPUnit 4.8.24 by Sebastian Bergmann and contributors.

Runtime:	PHP 7.0.3-13+deb.sury.org~trusty+1 with Xdebug 2.4.0RC4
Configuration:	/home/vagrant/Code/phpunit.xml

.............

Time: 12.21 seconds, Memory: 30.00Mb

OK (13 tests, 198 assertions)

**Comment by crynobone** [comment]
> Noone would go and copy paste new code change blindly lol. If they do, it's their problem.

The problem with changes on patch release is that when people upgrade from say `v5.2.0` (start the project when 5.2 released) to `v5.3.0`, often these between version changes are left out from upgrading guide. So yes, some including myself always update the skeleton. It helps us not to miss anything (and reduce errors when upgrading major version).

**Comment by shehi** [comment]
Why would you use a piece of code (in this case, the system's phpunit) 
over which you have zero control?! Even Xdebug messes up when something 
is outside working directory.

**Comment by shehi** [comment]
Agreed, that's why I recommended adding inline XML comments. I still am 
amazed why noone entertains the idea. Why XML's are any special? Does it 
break PHPUnit when you add a comment into phpunit.xml?

**Comment by StephenWattbike** [comment]
> Why would you use a piece of code (in this case, the system's phpunit)

is there something magically different between calling: phpunit from the project route or calling the file ./vendor/bin/phpunit ?

  xml comments work fine in the phpunit file.

**Comment by shehi** [comment]
It's hard to debug and control the flow as you debug, which is the 
inseparable part of Dev work. And last time I checked, PHPUnit was Dev 
tool. Otherwise, yes, they should be the same, unless whoever packaged 
it for, let's say Ubuntu, modified it to suit package-manager's needs. I 
never trust those packagers personally: had problem with Scala-via-Apt 
in past, had problem with Node-via-Apt, and it was nightmare for me to 
work via PHPUnit-via-Apt when it came to debugging with XDebug.

Also, wasn't it [!CDATA[ ... ]] blah blah the correct way of commenting 
in XML?! Or any PI (processing instruction) would do the trick, via 
<?ignore .... ?> (the same original reason why PHP had <?php ... ?> tag 
in first place). It's been while since I read XML-Bible, like 6 years 
now, so I might be wrong.

**Comment by StephenWattbike** [comment]
I actually don't know what the difference between running phpunit from my project root yet and running ./vendor/bin/phpunit. I just wanted to mention that one gives me an error without needing this extra phpunit.xml filtering and one doesn't. I have been commenting XML with    !CDATA is different and is a way of stopping execution within the block from my understanding.

**Comment by StephenWattbike** [comment]
vagrant@homestead:~/Code/vendor/bin$ ./phpunit --version
PHPUnit 4.8.24 by Sebastian Bergmann and contributors.

vagrant@homestead:~/Code$ phpunit --version
PHPUnit 5.3.0 by Sebastian Bergmann and contributors.

When I run phpunit from the project root its using my "system" phpunit
When I run ./vendor/bin/phpunit its using the phpunit inside the vendor folder not the other way round like you suggested earlier @shehi

**Comment by shehi** [comment]
I didn't suggest the other way around. Command "phpunit" uses 
system-wide-installed application, whereas "./vendor/bin/phpunit" is 
within your application boundaries. This shouldn't need explanation. 
What I meant, was that, running it from system-scope/namespace makes 
debugging a nightmare. Anyway, I don't even understand why we keep 
discussing this email after email :) like a freshman. Let's move on.

**Comment by driesvints** [comment]
I still seem to be getting this error. Excluded all stand-alone files in the exclude list. Running PHPUnit 4.8.24.

**Comment by derrekbertrand** [comment]
@driesvints Tested this on a fresh install of Laravel 5.2, and received this error when running the `phpunit` I installed globally using composer (`~/.composer/vendor/bin/phpunit`). As mentioned by @StephenWattbike , the error went away when running `./vendor/bin/phpunit`.

No idea why this works, but that's for smarter people than me to figure out.

**Comment by driesvints** [comment]
@derrekbertrand it doesn't really matter if you use global or local phpunit but it's a matter of which version you're using I believe. Can you run the following two commands and paste the output here?

```
phpunit --version
vendor/bin/phpunit --version
```

**Comment by derrekbertrand** [comment]
It may be the case, or it could not... Checking the version of the global binary yields 5.1, but actually running it - it thinks it is 4.8? That's the version Laravel installed! This is some chicanery, but nothing that actually has to do with Laravel.

As you can see, I can reproduce the error every time, however, phpunit may be lying about its version or misconfigured somehow.

Check this out:

    derrek@foxglove-0001 ~/projects/test $ phpunit --version
    PHPUnit 5.1.4 by Sebastian Bergmann and contributors.

    derrek@foxglove-0001 ~/projects/test $ vendor/bin/phpunit --version
    PHPUnit 4.8.26 by Sebastian Bergmann and contributors.

    derrek@foxglove-0001 ~/projects/test $ phpunit
    PHPUnit 4.8.26 by Sebastian Bergmann and contributors.

    .                                                                   1 / 1 (100%)

    Time: 101 ms, Memory: 10.00MB

    PHP Fatal error:  Uncaught Illuminate\Contracts\Container\BindingResolutionException: Target [Illuminate\Contracts\Debug\ExceptionHandler] is not instantiable. in /home/derrek/projects/test/vendor/laravel/framework/src/Illuminate/Container/Container.php:748
    ...DUMP...
    
    derrek@foxglove-0001 ~/projects/test $ vendor/bin/phpunit 
    PHPUnit 4.8.26 by Sebastian Bergmann and contributors.

    .

    Time: 99 ms, Memory: 10.00MB

    OK (1 test, 2 assertions)

**Comment by driesvints** [comment]
@derrekbertrand that's odd. Although you've confirmed that the problem happens when running at least PHPUnit `4.8.x`. Can you run `which phpunit` to see if you can find the path to the global binary? Afterwards can you try to run the 5.1.4 binary directly with the full path to see if it also happens with that particularly version?

**Comment by derrekbertrand** [comment]
@driesvints Breaks like you'd expect, and gets confused about its version.

    derrek@foxglove-0001 ~/projects/test $ which phpunit
    /home/derrek/.composer/vendor/bin/phpunit
    derrek@foxglove-0001 ~/projects/test $ /home/derrek/.composer/vendor/bin/phpunit
    PHPUnit 4.8.26 by Sebastian Bergmann and contributors.

    .                                                                   1 / 1 (100%)

    Time: 102 ms, Memory: 10.00MB

    PHP Fatal error:  Uncaught Illuminate\Contracts\Container\BindingResolutionException: Target [Illuminate\Contracts\Debug\ExceptionHandler] is not instantiable. in /home/derrek/projects/test/vendor/laravel/framework/src/Illuminate/Container/Container.php:748
    ...DUMP...

**Comment by driesvints** [comment]
@derrekbertrand super weird :-| Don't know what's going on here.

**Comment by mihailomisic** [comment]
I had the same issue. _(Using Laravel 5.2)_
The problem was that i had a newer version of **phpunit** installed globally (5.4.4).
So by running **`phpunit`** command the global one was used.

Solutions:
1. Use **`./vendor/bin/phpunit`** instead;
2. Downgrade the global **phpunit**, after which you can use the **`phpunit`** command regularly.
```
composer global remove phpunit/phpunit
composer global require 'phpunit/phpunit=~4.0'
```
Hope somebody finds this useful.

**Comment by GrahamCampbell** [comment]
Or just don't globally require PHPUnit.

**Comment by jpuck** [comment]
@GrahamCampbell Is this no longer an issue because it was reverted in 288e361affbb1fb3b5ea3b8ad0c74baff227ed4f? I would have instead expected the path to be updated to something like:

```xml
<exclude>
    <directory>./routes</directory>
</exclude>
```

I only ask because I'm trying to investigate why [code coverage reports are throwing exceptions again in php 7.2 for laravel tests][1]. It doesn't seem to be related to the routes files, but I just thought I'd clarify this particular issue along the way.

Thanks

[1]:https://stackoverflow.com/q/44550410/4233593

**Comment by FatBoyXPC** [comment]
The routes directory is outside of `app` now (currently in L5.5, at least) but I'm still having this problem trying to run code coverage.

```
➜  matchdevs git:(master) ✗ vendor/bin/phpunit --coverage-html coverage
PHPUnit 6.4.3 by Sebastian Bergmann and contributors.

........................                                          24 / 24 (100%)

Time: 3.17 seconds, Memory: 26.00MB

OK (24 tests, 67 assertions)

Generating code coverage report in HTML format ...PHP Fatal error:  Uncaught Illuminate\Contracts\Container\BindingResolutionException: Target [Illuminate\Contracts\Debug\ExceptionHandler] is not instantiable. in /home/james/dev/matchdevs/vendor/laravel/framework/src/Illuminate/Container/Container.php:918
Stack trace:
#0 /home/james/dev/matchdevs/vendor/laravel/framework/src/Illuminate/Container/Container.php(758): Illuminate\Container\Container->notInstantiable('Illuminate\\Cont...')
#1 /home/james/dev/matchdevs/vendor/laravel/framework/src/Illuminate/Container/Container.php(631): Illuminate\Container\Container->build('Illuminate\\Cont...')
#2 /home/james/dev/matchdevs/vendor/laravel/framework/src/Illuminate/Container/Container.php(586): Illuminate\Container\Container->resolve('Illuminate\\Cont...', Array)
#3 /home/james/dev/matchdevs/vendor/laravel/framework/src/Illuminate/Foundation/Application.php(732): Illuminate\Container\Container->make('Illuminate\\Cont...', Array)
#4 /home/james/dev/matchdevs/vendor/laravel/framework/src/Illuminate/Foundation/Bootstrap/HandleExceptio in /home/james/dev/matchdevs/vendor/laravel/framework/src/Illuminate/Container/Container.php on line 918
PHP Fatal error:  Uncaught Illuminate\Contracts\Container\BindingResolutionException: Target [Illuminate\Contracts\Debug\ExceptionHandler] is not instantiable. in /home/james/dev/matchdevs/vendor/laravel/framework/src/Illuminate/Container/Container.php:918
Stack trace:
#0 /home/james/dev/matchdevs/vendor/laravel/framework/src/Illuminate/Container/Container.php(758): Illuminate\Container\Container->notInstantiable('Illuminate\\Cont...')
#1 /home/james/dev/matchdevs/vendor/laravel/framework/src/Illuminate/Container/Container.php(631): Illuminate\Container\Container->build('Illuminate\\Cont...')
#2 /home/james/dev/matchdevs/vendor/laravel/framework/src/Illuminate/Container/Container.php(586): Illuminate\Container\Container->resolve('Illuminate\\Cont...', Array)
#3 /home/james/dev/matchdevs/vendor/laravel/framework/src/Illuminate/Foundation/Application.php(732): Illuminate\Container\Container->make('Illuminate\\Cont...', Array)
#4 /home/james/dev/matchdevs/vendor/laravel/framework/src/Illuminate/Foundation/Bootstrap/HandleExceptio in /home/james/dev/matchdevs/vendor/laravel/framework/src/Illuminate/Container/Container.php on line 918
➜  matchdevs git:(master) ✗ 
```

**Pull request merged**